networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/RateTracker.java

```diff
@@ -35,27 +39,108 @@
   // they can have the objects they request otherwise they get none.
-  public synchronized long popObjectRequests(final long objectCount) {
+  public synchronized Optional<ObjectsRequestResponse> popObjectRequests(final long objectCount) {
     pruneRequests();
+    final UInt64 currentTime = timeProvider.getTimeInSeconds();
     if ((peerRateLimit - requestsWithinWindow) <= 0) {
-      return 0L;
+      return Optional.empty();
     }
 
     requestsWithinWindow += objectCount;
-    requestCount.compute(
-        timeProvider.getTimeInSeconds(),
-        (key, val) -> val == null ? objectCount : val + objectCount);
-    return objectCount;
+    final ObjectsRequestResponse objectsRequestResponse =
+        new ObjectsRequestResponse.ObjectsRequestBuilder(objectCount)
+            .timeSeconds(currentTime)
+            .build();
+    requestCount.put(new ObjectRequestsEntryKey(objectsRequestResponse), objectCount);
+    return Optional.of(objectsRequestResponse);
+  }
+
+  public synchronized void adjustObjectRequests(
+      final ObjectsRequestResponse objectsRequestResponse, final long returnedObjectsCount) {
+    pruneRequests();
+    final ObjectRequestsEntryKey objectRequestsEntryKey =
+        new ObjectRequestsEntryKey(objectsRequestResponse);
+    if (requestCount.containsKey(objectRequestsEntryKey)) {
+      final long initialObjectsCount = requestCount.get(objectRequestsEntryKey);
+      requestCount.put(objectRequestsEntryKey, returnedObjectsCount);
+      requestsWithinWindow = requestsWithinWindow - initialObjectsCount + returnedObjectsCount;
+    }
   }
 
   void pruneRequests() {
     final UInt64 currentTime = timeProvider.getTimeInSeconds();
     if (currentTime.isLessThan(timeoutSeconds)) {
       return;
     }
-    final NavigableMap<UInt64, Long> headMap =
-        requestCount.headMap(currentTime.minus(timeoutSeconds), false);
+    final NavigableMap<ObjectRequestsEntryKey, Long> headMap =
+        requestCount.headMap(
+            new ObjectRequestsEntryKey(currentTime.minus(timeoutSeconds), UUID.randomUUID()),
+            false);
     headMap.values().forEach(value -> requestsWithinWindow -= value);
     headMap.clear();
   }
+
+  public static class ObjectRequestsEntryKey implements Comparable<ObjectRequestsEntryKey> {
+    private final UInt64 timeSeconds;
+    private final UUID requestId;
+
+    public ObjectRequestsEntryKey(final UInt64 timeSeconds, final UUID requestId) {
+      this.timeSeconds = timeSeconds;
+      this.requestId = requestId;
+    }
+
+    public ObjectRequestsEntryKey(final ObjectsRequestResponse objectsRequestResponse) {
+      this.timeSeconds = objectsRequestResponse.timeSeconds;
+      this.requestId = objectsRequestResponse.requestId;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(timeSeconds, requestId);
+    }
+
+    @Override
+    public int compareTo(@NotNull ObjectRequestsEntryKey other) {
+      return timeSeconds.compareTo(other.timeSeconds);
```

Comment:
Actually I didn't really forget, since we use `headMap` to retrieve sub maps based on time only, I explicitly ignored the request ids... Added it now (even though it's not useful for the current implementation, that wouldn't be aligned with `compareTo` definition)